libcxx/include/__memory/start_lifetime_as_array.h

```diff
@@ -0,0 +1,47 @@
+// -*- C++ -*-
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _LIBCPP___MEMORY_START_LIFETIME_AS_ARRAY_H
+#define _LIBCPP___MEMORY_START_LIFETIME_AS_ARRAY_H
+
+#include <__config>
+#include <__memory/start_lifetime_as.h>
+#include <cstddef>
+
+#if !defined(_LIBCPP_HAS_NO_PRAGMA_SYSTEM_HEADER)
+#  pragma GCC system_header
+#endif
+
+_LIBCPP_BEGIN_NAMESPACE_STD
+
+template <class _Tp>
+_LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR _Tp* start_lifetime_as_array(void* __p, [[maybe_unused]] size_t __n) _NOEXCEPT {
+  return static_cast<_Tp*>(__p);
```

Comment:
No built-in used here. Could you explain why this is considered to work?

Reply:
The function itself doesn't actually do anything at runtime.  It's a no-op in terms of generated code. It happens at the compiler level. When the compiler sees a call to this function, it treats it as a signal that it should consider the lifetime of an array of  `_Tp` objects to have begun at the memory location pointed to by `__p` , even though no constructors have been called. Or is it intended something else? I am not sure, Could you explain?

Reply:
> When the compiler sees a call to this function,

Currently, I'm not seeing anything in the source code of Clang or GCC that recognizes `start_lifetime_as(_array)`. So I can't believe this approach will automatically work,

> Or is it intended something else? I am not sure, Could you explain?

I _guess_ an intrinsic like `__builitin_start_lifetime_as` (or `__builtin_tbaa_barrier`?) is expected.

Reply:
Ah Okay, that make sense, is it worth it if we currently define a macro to check if the compiler supports the necessary intrinsics?

If supported, it declares the intrinsics and implements `start_lifetime_as` and `start_lifetime_as_array` using these intrinsics. If not supported, it falls back to the current implementation (which is essentially a no-op).

Reply:
I think that's a bad idea, since people expect things to work as advertised that exist. IOW, these functions should only exist if we can actually implement them.